modules/bootstrap/main.tf

```diff
@@ -61,47 +40,109 @@ locals {
+  quota                = coalesce(each.value.quota, var.file_shares_configuration.quota)
+  access_tier          = coalesce(each.value.access_tier, var.file_shares_configuration.access_tier)
+}
+
+# https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/data-sources/storage_share
+data "azurerm_storage_share" "this" {
+  for_each = var.file_shares_configuration.create_file_shares ? {} : var.file_shares
+
+  name                 = each.value.name
   storage_account_name = local.storage_account.name
-  quota                = var.storage_share_quota
-  access_tier          = var.storage_share_access_tier
 
   lifecycle {
     precondition {
-      condition = var.storage_share_name != null ? alltrue([
-        can(regex("^[a-z0-9](-?[a-z0-9])+$", var.storage_share_name)),
-        can(regex("^([a-z0-9-]){3,63}$", var.storage_share_name))
-      ]) : true
-      error_message = "A File Share name must be between 3 and 63 characters, all lowercase numbers, letters or a dash, it must follow a valid URL schema."
+      condition     = !var.file_shares_configuration.create_file_shares && !var.create_storage_account
+      error_message = "You cannot source File Shares from a newly created Storage Account."
     }
   }
 }
 
+locals {
+  file_shares     = var.file_shares_configuration.create_file_shares ? azurerm_storage_share.this : data.azurerm_storage_share.this
+  package_folders = ["content", "config", "software", "plugins", "license"]
+}
+
+# https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/storage_share_directory
 resource "azurerm_storage_share_directory" "this" {
-  for_each = var.storage_share_name != null ? toset([
-    "content",
-    "config",
-    "software",
-    "plugins",
-    "license"
-  ]) : toset([])
-
-  name                 = each.key
-  share_name           = azurerm_storage_share.this[0].name
+  for_each = {
+    for v in setproduct(keys(var.file_shares), local.package_folders) :
+    join("-", v) => {
+      share_key   = v[0]
+      folder_name = v[1]
+    }
+    if !var.file_shares_configuration.disable_package_dirs_creation
+  }
+
+  name                 = each.value.folder_name
+  share_name           = local.file_shares[each.value.share_key].name
   storage_account_name = local.storage_account.name
 }
 
+
+locals {
+  bootstrap_filenames = {
+    for k, v in var.file_shares : k => {
+      for f in try(fileset(v.bootstrap_package_path, "**"), {}) : f => "${v.bootstrap_package_path}/${f}"
+    }
+  }
+
+  # invert var.files map 
+  inverted_files = {
+    for k, v in var.file_shares : k => {
+      for k, v in v.bootstrap_files : v => k
+    }
+  }
+
+  inverted_filenames = {
+    for k, _ in var.file_shares : k => merge(local.bootstrap_filenames[k], local.inverted_files[k])
+  }
+
+  # invert local.filenames map
+  filenames = {
+    for k, _ in var.file_shares : k => { for _k, _v in local.inverted_filenames[k] : _v => _k }
+  }
+  filenames_across_fileshares_flat = flatten([
```

Comment:
Can we in comments describe in few sentences what is the purpose of having inverted files , then flat filenames across file shares etc. ? It's not clear while looking on that code at the beginning.